Improve Path Pattern Expression documentation (#723)
neo4j/docs-cypher: modules/ROOT/pages/clauses/where.adoc

2+|Rows: 1
|===

[[query-where-patterns]]
== Path pattern expressions

> [[path-pattern-expressions]]

Reviewer: This anchor should be placed above the header (and replace the old one (query-where-patterns), which ( checked) is currently not used).

```suggestion
```

2+|Rows: 1
|===

[[query-where-patterns]]
== Path pattern expressions

[[path-pattern-expressions]]

> === About path pattern expressions

Reviewer: ```suggestion
```

Path pattern expressions can be used to assert whether a specified path exists at least once in a graph. Path pattern expressions are similar to xref::subqueries/existential.adoc[existential subqueries].

> On one hand, xref::subqueries/existential.adoc[existential subqueries] are the more powerful feature and can do anything path pattern expressions can, while on the other hand path pattern expressions are more concise.

Reviewer: ```suggestion
Similar to xref::subqueries/existential.adoc[existential subqueries], path pattern expressions can be used to assert whether a specified path exists at least once in a graph.
While existential subqueries are more powerful and capable of performing anything achievable with path pattern expressions, path pattern expressions are more concise.
```

Path pattern expressions come with several restrictions which are highlighted below.

> Use cases which require more permissive functionality should consider using xref::subqueries/existential.adoc[existential subqueries] instead.

Reviewer: ```suggestion
Path pattern expressions have the following restrictions (use cases that require extended functionality should consider using xref::subqueries/existential.adoc[existential subqueries] instead):
```

> Path pattern expressions may only use a subset of xref::patterns/reference.adoc#graph-patterns[graph patterns] semantics.

Reviewer: ```suggestion
* Path pattern expressions may only use a subset of xref::patterns/reference.adoc#graph-patterns[graph pattern] semantics.
```

A path pattern expressions must be a xref::patterns/reference.adoc#path-patterns[path pattern] of length greater than zero.

> In other words, it must contain at least one xref::patterns/reference.adoc#relationship-patterns[relationship] or xref::patterns/reference.adoc#variable-length-relationships[variable-length relationship].

Reviewer: ```suggestion
* A path pattern expression must be a xref::patterns/reference.adoc#path-patterns[path pattern] of length greater than zero.
In other words, it must contain at least one xref::patterns/reference.adoc#relationship-patterns[relationship] or xref::patterns/reference.adoc#variable-length-relationships[variable-length relationship].
```

Should we not link to quantified relationships rather than variable-length relationships (https://neo4j.com/docs/cypher-manual/current/patterns/reference/#quantified-relationships)? With the release of qpps, it was my understanding that we should not promote the old var-length syntax, but I may be missing something here.

Author: I could maybe remove the part where I mention the old var-length, if we didn't want to reference that at all? Unfortunately, path pattern expressions cannot use quantified relationships or quantified path patterns.

Reviewer: I see. In that case, I think it is better to leave var-lengths in.

Path pattern expressions may not declare new variables.
They can only reference existing variables.

Path pattern expressions may only be used in positions where a xref:queries/expressions.adoc#boolean[boolean expression] is expected.

> The following sections will demonstrate how to use path pattern expressions in a `WHERE` clause.

Reviewer: ```suggestion
* Path pattern expressions may not declare new variables.
They can only reference existing variables.

* Path pattern expressions may only be used in positions where a xref:queries/expressions.adoc#boolean[boolean expression] is expected.
The following sections will demonstrate how to use path pattern expressions in a `WHERE` clause.
```

[[filter-on-patterns]]

> === Filter on patterns

Reviewer: I think this needs a more distinct header (I have added the "Filter on patterns" as a suggestion below (line 509)

```suggestion
=== Note on path patterns and path pattern expressions
```

Author: I went back and forth a bit, and settled on "About path pattern expressions"

> Path patterns may be used both in a MATCH clause and in a WHERE clause. When used in a WHERE clause they are called path

Reviewer 2: There are more positions for path patterns than just the `WHERE` clause. E.g.:
- `exists(<path pattern>)`
- `CASE WHEN <path pattern> THEN ... END`

They in turn can be used in all places that accept expressions.

> pattern expressions.

Reviewer: ```suggestion
xref:patterns/concepts.adoc#path-patterns[Path patterns] may be used both in a `MATCH` clause and in a `WHERE` clause.
When used in a `WHERE` clause they are called path pattern expressions.
```

Path patterns used in a MATCH clause should generally be thought of as retrieving data. Path patterns expressions used

> in a WHERE clause should on the other hand be thought of as eliminating data. The clause `MATCH (a)-[]->(b)` is very

Reviewer 2: I find "eliminating" sounds a lot like we are `DELETE`ing. The purpose of `WHERE` in general is to _filter_  the results/data that we found with `MATCH`.

What do you think about contrasting retrieving (or maybe "enumerating") in `MATCH` with "testing"  in `WHERE`? One could probably also use/promote the `EXISTS { ... }` or `exists(...)` expression here to explain what the pattern does in `WHERE`.

Author: I agree that "eliminating" was a poor choice of words. I think that "retrieving" and "filtering" are good though. The page already uses "filtering" in many places which means the user should have a grasp of the concept by the time he reaches this section.

Author: Both Jens and you have offered some suggestions about linking to other parts of the documentation which have their own sections in the docs. It would feel a bit spaghetti to me to include something about those other sections in here, so I may extract the "path pattern expressions" to their section of the docs and then link to them from relevant places such as WHERE (this page), CASE WHEN (other page), exists function (other page). I would not mention the EXISTS subquery as that has no special handling of path pattern expressions - if the EXISTS subquery has a WHERE clause then it can accept a path pattern expression, but so can anything else with a WHERE clause.

Author: I feel that using `EXISTS { p }` or `exists( p )` in here will be quite confusing for users. This is because `EXISTS { }` accepts a path pattern whereas `exists( )` accepts has a parameter which is not a boolean but works with a pattern expression anyways (this is the only exception to the rule about pattern expressions). 

Given that ultimately we want to get rid of `exists( p )` altogether, I was hoping to not have to dive into the exception and KISS instead. I find the exception a bit superfluous anyways (YMMV) because `exists( pattern-expression )` is semantically the same as simply `pattern-expression`. They are both an existence check. 

As far as I can see, the purpose of `exists( pattern-expression )` is either as syntactic sugar or to allow using a pattern expression in places where any expression can be used.

Reviewer 2: I like "filtering". That goes well together with the rest of the text.
What I meant with the `EXISTS { ... }` expression is that `WHERE (a)-[]→(b)` can be thought of as a shorthand of `WHERE EXISTS { (a)-[]→(b) }`.

I guess, pattern expressions are independent enough from `WHERE` clauses that they may reside in a page of their own. But @JPryce-Aklundh probably has an opinion on extracting that section.

Reviewer 2: You are right, `exists ( pattern-expression )` is nothing we want to promote. We mostly keep it as a legacy and to not unnecessarily introduce breaking changes.

I am not sure I understand your point about `EXISTS { ... }`, though: `EXISTS { ... }` is the more powerful construct and can accept arbitrary queries, including what can be expressed as path pattern expression. My idea was to frame path pattern expression as a short-hand to this. That way we promote `EXISTS { ... }` as a means when people want to express more than what can be expressed in a path pattern expression. And if people know `EXISTS { ... }` already, then such an equivalence will make it easier for them to learn path pattern expressions.

Reviewer 2: This is just an idea, of course, so feel free to ignore it.

Author: > I guess, pattern expressions are independent enough from WHERE clauses that they may reside in a page of their own. But @JPryce-Aklundh probably has an opinion on extracting that section.

I am going to try to make it work by keeping it on the current page. Other options I considered was moving it to the ["Cypher Expressions"](https://neo4j.com/docs/cypher-manual/current/queries/expressions/) page, or somewhere in the ["Cypher Patterns"](https://neo4j.com/docs/cypher-manual/current/patterns/) pages. All the options have pros and cons. I am finding it hard to get a clean result, but hopefully we can reach something that's good enough - certainly something that is better than what we had before.

different from `WHERE (a)-[]->(b)`. The first will produce a record for every path it can find between `a` and `b`,

> whereas the latter will eliminate records where `a` and `b` do not have a directed relationship between them.

Reviewer: ```suggestion
Path patterns used in a `MATCH` clause should be thought of as retrieving data.
Path pattern expressions used in a `WHERE` clause should be thought of as eliminating data.

For example, the clause `MATCH (a)-[]->(b)` is very different from `WHERE (a)-[]->(b)`.
The first will produce a record for every path it can find between `a` and `b`.
The latter will eliminate records where `a` and `b` do not have a directed relationship between them.
```

> There are several limitations for what path patterns can do in contrast to their path pattern counterparts.

Reviewer: ```suggestion
There are several limitations to what path patterns used in a `MATCH` clause can do in contrast to path pattern expressions used in a `WHERE` clause.
```

Reviewer: I'm thinking the below information might be more efficiently communicated in a two column table (path patterns/ path pattern expressions) - what do you think?

Reviewer 2: I think you have a point @JPryce-Aklundh, but I understood the two exactly opposite:
```suggestion
There are several limitations to what path patterns expressions used in a `WHERE` clause can do in contrast to path pattern used in a `MATCH` clause.
```

> - Path patterns in a MATCH clause can introduce new variables. Path pattern expressions in a WHERE clause cannot.

Reviewer: ```suggestion
- Path patterns in a `MATCH` clause can introduce new variables.
The same is not true for path pattern expressions in a `WHERE` clause.
```

- Path patterns in a MATCH clause can enjoy the full range of xref::patterns/reference.adoc[Pattern] semantics. Path
pattern expressions in a WHERE clause on the other hand can only use
xref::patterns/reference.adoc#relationship-patterns[Relationships] and

> xref::patterns/reference.adoc#variable-length-relationships[Variable-length relationships] semantics.

Reviewer: ```suggestion
- All  xref::patterns/reference.adoc[pattern] semantics apply to path patterns in a `MATCH` clause.
Path pattern expressions in a `WHERE` clause can only use
xref::patterns/reference.adoc#relationship-patterns[relationship] and
xref::patterns/reference.adoc#variable-length-relationships[variable-length relationship] semantics.
```

- Multiple path patterns in a MATCH clause can be comma separated. Multiple path pattern expressions in a WHERE clause

> can only be separated by xref::syntax/operators.adoc#query-operators-boolean[Boolean operators] such as `AND` and `OR`.

Reviewer: ```suggestion
- Multiple path patterns in a `MATCH` clause can be separated by commas.
Multiple path pattern expressions in a `WHERE` clause can only be separated by xref::syntax/operators.adoc#query-operators-boolean[Boolean operators] such as `AND` and `OR`.

=== Filter on patterns
```

.Query
[source, cypher]
----
MATCH (n:Person)
WHERE (n)-[:KNOWS]-({name: 'Timothy'})
RETURN n.name, n.age
MATCH (other:Person)
WHERE (other)-[:KNOWS]-({name: 'Timothy'})
RETURN other.name, other.age
----

The `name` and `age` values are returned for nodes that have a relationship with the type `KNOWS` connected to `Timothy`:

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| n.name | n.age
| other.name | other.age
| "Andy" | 36
2+|Rows: 1
|===

[[filter-on-relationship-type]]
=== Filter on relationship type

You can put the exact relationship type in the `MATCH` pattern,
It also possible to do more advanced filtering on the relationship type by using the special property `type` to compare the type with something else.
In this example, the query does a regular expression comparison with the name of the relationship type:

.Query
[source, cypher]
----
MATCH (n:Person)-[r]->()
WHERE n.name='Andy' AND type(r) =~ 'K.*'
RETURN type(r), r.since
----

This returns all relationships having a type whose name starts with 'K':

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| type(r) | r.since
| "KNOWS" | 1999
| "KNOWS" | 2012
2+|Rows: 2
|===

> 

Author: This did not belong in this section, as it has nothing to do with path pattern expressions. 

If we need a new section to discuss the `type` function and its siblings `types`, `label`, and `labels`, then I would suggest making a new card for them. There however is already documentation concerning these on the https://neo4j.com/docs/cypher-manual/current/functions/scalar page.
